eclass/dkms.eclass

```diff
@@ -0,0 +1,527 @@
+# the current working directory is used.
+#
+# For convenience this eclass exports a src_compile function that runs
+# dkms_autoconf if the dkms USE flag is enabled, and if the flag is
+# disabled it runs linux-mod-r1_src_compile instead. Similarly,
+# the src_install function exported by this eclass finds any
+# dkms.conf files in the current working directory or one of its
+# subdirectories and then calls dkms_dopackage for these packages.
+# And if the dkms USE flag is disabled it runs
+# linux-mod-r1_src_install instead.
+#
+# The pkg_postinst and pkg_postrm functions then take care of
+# (de)registering, (un)building, removing, and/or adding the DKMS
+# packages. For convenience the eclass also exports a pkg_config
+# function that rebuilds and reinstalls any DKMS packages the ebuild
+# owns for the currently running kernel.
+#
+# @EXAMPLE:
+#
+# To add DKMS support to an ebuild currently using only linux-mod-r1.
+#
+# Change:
+#
+# @CODE
+# inherit linux-mod-r1
+#
+# src_compile() {
+#     local modlist=(
+#         gentoo
+#         gamepad=kernel/drivers/hid:gamepad:gamepad/obj
+#     )
+#     local modargs=( NIH_SOURCE="${KV_OUT_DIR}" )
+#
+#     linux-mod-r1_src_compile
+# }
+# @CODE
+#
+# To:
+#
+# @CODE
+# inherit dkms
+#
+# src_compile() {
+#     local modlist=(
+#         gentoo
+#         gamepad=kernel/drivers/hid:gamepad:gamepad/obj
+#     )
+#     local modargs=( NIH_SOURCE="${KV_OUT_DIR}" )
+#
+#     dkms_src_compile
+# }
+# @CODE
+#
+# Note that due to the inherit order the src_install and pkg_postinst
+# phase functions may have to be defined explicitly.
+#
+# @EXAMPLE:
+#
+# A more complex example is the case of an ebuild that is currently
+# inheriting linux-mod-r1, but is not using any of its phase
+# functions. In this case there is usually no modlist for
+# dkms_autoconf to convert into a DKMS configuration file.
+# Instead the ebuild must utilize a dkms.conf provided by upstream
+# in the sources, or alternatively create one from scratch and
+# include it in FILESDIR.
+#
+# Tip: Check if there is a rpm/deb spec or similar script that can
+# create a dkms.conf to find a hint of what it should look like and
+# where it should be created for this particular package.
+#
+# @CODE
+# inherit dkms linux-mod-r1
+#
+# src_prepare() {
+#   default
+#   sed -e "s/@VERSION@/${PV}/" -i modules/dkms.conf || die
+# }
+#
+# src_compile() {
+#     if use dkms; then
+#         dkms_gentoofy_conf modules/dkms.conf
+#     else
+#         emake "${MODULES_MAKEARGS[@]}" modules
+#     fi
+# }
+#
+# src_install() {
+#     if use dkms; then
+#         dkms_dopackage modules
+#     else
+#         linux_domodule modules/mymodule.ko
+#         modules_post_process
+#     fi
+#     einstalldocs
+# }
+#
+# pkg_postinst() {
+#     dkms_pkg_postinst
+# }
+# @CODE
+
+case ${EAPI} in
+	8) ;;
+	*) die "${ECLASS}: EAPI ${EAPI:-0} not supported" ;;
+esac
+
+if [[ -z ${_DKMS_ECLASS} ]]; then
+_DKMS_ECLASS=1
+
+inherit linux-mod-r1
```

Comment:
I appreciate that this is the natural solution if Ionen isn't sure (or rejects it) for linux-mod-r1, but having basically every ebuild change from `linux-mod-r1` to `kmod` for what it inherits instead feels like essentially the same thing.

Reply:
> instead feels like essentially the same thing.

I suppose. But a fundamental difference with my first attempt is that now we're not adding `USE=dkms` to everything blindly, which was a big flaw in my original design where this was all part of `linux-mod-r1`. I actually tested this now on each package where the new flag is added and verified that it works.